File: scripts/open_research_games_portal/generate_game_pages.R
#===============================================================================
# Generate Individual Game Pages Script
#===============================================================================
#
# Purpose: Creates individual Hugo content pages for each game in the portal
#          This improves SEO, enables direct linking, and avoids URL parameter issues
#
# Input: data/open_research_games.json
# Output: content/open-research-games-portal/games/{game-slug}/index.md files
#
# Usage: Rscript scripts/open_research_games_portal/generate_game_pages.R
#
#===============================================================================

library(jsonlite)
library(stringr)

# Read the games data
games_data <- jsonlite::fromJSON("data/open_research_games.json")

# Create the games directory if it doesn't exist
games_dir <- "content/open-research-games-portal/games"
if (!dir.exists(games_dir)) {
  dir.create(games_dir, recursive = TRUE)
}

# Function to clean text for YAML frontmatter
clean_for_yaml <- function(text) {
  if (is.null(text) || text == "" || is.na(text)) {
    return("")
  }
  # Escape quotes and special YAML characters
  text <- gsub('"', '\\"', text)
  text <- gsub("'", "\\'", text)
  text <- gsub('\n', ' ', text)
  text <- gsub('\r', '', text)
  text <- gsub(':', '\\:', text)
  return(text)
}

# Function to create frontmatter
create_frontmatter <- function(game_slug, game_data) {
  title <- clean_for_yaml(game_data$title %||% "Untitled Game")
  description <- clean_for_yaml(game_data$description %||% "")
  
  # Truncate description for meta description (max 160 chars)
  meta_description <- if (nchar(description) > 160) {
    paste0(substr(description, 1, 157), "...")
  } else {
    description
  }
  
  frontmatter <- paste0(
    "---\n",
    "title: |\n  ", title, "\n",
    "type: \"game-detail\"\n",
    "layout: \"single\"\n",
    "description: |\n  ", meta_description, "\n",
    "game_slug: \"", game_slug, "\"\n",
    "---\n\n"
  )
  
  return(frontmatter)
}

# Generate pages for each game
cat("Generating individual game pages...\n")
generated_count <- 0

for (game_slug in names(games_data)) {
  game_data <- games_data[[game_slug]]
  
  # Create directory for this game
  game_page_dir <- file.path(games_dir, game_slug)
  if (!dir.exists(game_page_dir)) {
    dir.create(game_page_dir, recursive = TRUE)
  }
  
  # Create the index.md file
  index_file <- file.path(game_page_dir, "index.md")
  
  # Generate content
  frontmatter <- create_frontmatter(game_slug, game_data)
  content <- paste0(
    frontmatter,
    "{{< single-game-details >}}\n"
  )
  
  # Write the file
  writeLines(content, index_file)
  generated_count <- generated_count + 1
}

cat("Generated", generated_count, "individual game pages\n")
cat("Pages created in:", games_dir, "\n")
cat("URL format: /open-research-games-portal/games/{game-slug}/\n")

File: scripts/open_research_games_portal/generate-game-pages.py
#!/usr/bin/env python3
"""
Open Research Games Portal - Game Pages Generator

Creates individual Hugo pages for each game with server-side rendered content.
No JavaScript required - everything is generated at build time.

Usage: python3 scripts/open_research_games_portal/generate-game-pages.py
"""

import json
import os
import shutil
from pathlib import Path

def clean_for_yaml(text):
    """Clean text for YAML frontmatter"""
    if not text:
        return ""
    return str(text).replace('"', '\\"').replace('\n', ' ').replace('\r', '')

def create_game_page(game_slug, game_data, output_dir):
    """Create individual game page"""
    game_dir = output_dir / game_slug
    game_dir.mkdir(parents=True, exist_ok=True)
    
    title = game_data.get('title', 'Untitled Game')
    description = game_data.get('description', '')
    
    # Truncate description for meta
    if len(description) > 160:
        description = description[:157] + "..."
    
    # Create frontmatter
    frontmatter = f'''---
title: "{clean_for_yaml(title)}"
type: "open-research-games-portal"
layout: "single"
description: "{clean_for_yaml(description)}"
url: "/games/{game_slug}/"
date: "2025-01-01"
draft: false
---

'''
    
    # Use Hugo shortcode instead of raw HTML to avoid escaping issues
    content = frontmatter + f'{{{{< game-details-static slug="{game_slug}" >}}}}\n'
    
    # Write the file
    index_file = game_dir / "index.md"
    with open(index_file, 'w', encoding='utf-8') as f:
        f.write(content)

def main():
    # Load games data
    with open('data/open_research_games.json', 'r', encoding='utf-8') as f:
        games_data = json.load(f)
    
    # Create output directory
    output_dir = Path('content/games_details')
    if output_dir.exists():
        shutil.rmtree(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    
    # Create individual pages
    for game_slug, game_data in games_data.items():
        create_game_page(game_slug, game_data, output_dir)
    
    print(f"✅ Created {len(games_data)} individual game pages")
    print(f"URLs like: /games/copyright-the-card-game-id-034/")

if __name__ == "__main__":
    main()
